crates/extensions/tedge_mqtt_ext/src/lib.rs

```diff
@@ -0,0 +1,207 @@
+    async fn spawn(self, runtime: &mut RuntimeHandle) -> Result<(), RuntimeError> {
+        let (mqtt_actor, mqtt_message_box) = self.build().await;
+
+        runtime.run(mqtt_actor, mqtt_message_box).await?;
+        Ok(())
+    }
+}
+
+struct MqttMessageBox {
+    mqtt_client: mqtt_channel::Connection,
+    peer_receiver: Receiver<MqttMessage>,
+    peer_senders: Vec<(TopicFilter, DynSender<MqttMessage>)>,
+}
+
+impl MqttMessageBox {
+    async fn new(
+        mqtt_config: mqtt_channel::Config,
+        peer_receiver: Receiver<MqttMessage>,
+        peer_senders: Vec<(TopicFilter, DynSender<MqttMessage>)>,
+    ) -> Result<Self, MqttError> {
+        let mqtt_client = mqtt_channel::Connection::new(&mqtt_config).await?;
+        Ok(MqttMessageBox {
+            mqtt_client,
+            peer_receiver,
+            peer_senders,
+        })
+    }
+}
+
+#[async_trait]
+impl MessageBox for MqttMessageBox {
+    type Input = MqttActorMessage;
+    type Output = MqttActorMessage;
+
+    async fn recv(&mut self) -> Option<MqttActorMessage> {
+        tokio::select! {
+            Some(message) = self.peer_receiver.next() => {
+                let message = MqttActorMessage::PeerMessage(message);
+                self.log_input(&message);
+                Some(message)
+            },
+            Some(message) = self.mqtt_client.received.next() => {
+                Some(MqttActorMessage::DirectMessage(message))
+            },
+            else => None,
+        }
+    }
+
+    async fn send(&mut self, message: MqttActorMessage) -> Result<(), ChannelError> {
+        match message {
+            MqttActorMessage::DirectMessage(message) => {
+                // FIXME one should trace all the sent copies of this message.
+                self.log_output(&MqttActorMessage::DirectMessage(message.clone()));
+
+                for (topic_filter, peer_sender) in self.peer_senders.iter_mut() {
+                    if topic_filter.accept(&message) {
+                        let message = message.clone();
+                        peer_sender.send(message).await?;
+                    }
+                }
+            }
+            MqttActorMessage::PeerMessage(message) => {
+                self.mqtt_client
+                    .published
+                    .send(message)
+                    .await
+                    .expect("TODO catch actor specific errors");
+            }
+        }
+        Ok(())
+    }
+
+    fn turn_logging_on(&mut self, _on: bool) {}
+
+    fn name(&self) -> &str {
+        "MQTT"
+    }
+
+    fn logging_is_on(&self) -> bool {
+        true
+    }
+}
+
+struct MqttActor;
+
+#[async_trait]
+impl Actor for MqttActor {
+    type MessageBox = MqttMessageBox;
+
+    fn name(&self) -> &str {
+        "MQTT"
+    }
+
+    async fn run(mut self, mut mailbox: MqttMessageBox) -> Result<(), ChannelError> {
+        while let Some(message) = mailbox.recv().await {
+            mailbox.send(message).await?;
+        }
```

Comment:
Without a look to the message box implementation, one has the incorrect impression that this actor is a toy echo actor!

This is yet another reason to move the code dealing with the MQTT connection out of the mail box.

```suggestion
        loop {
            tokio::select! {
                Some(message) = self.mailbox.next() => {
                    self.mqtt_client.published.send(message).await?
                },
                Some(message) = self.mqtt_client.received.next() => {
                    self.mailbox.send(message).await?
                },
                else => return Ok(),               
            }
        }
```